index.src.html

```diff
@@ -2394,14 +2420,692 @@
+        `-1`.
+    3. Let |encodedFrame| be the encoded frame identified by |frameIndex| and
+        {{ImageDecoder/[[internal selected track index]]}}.
+    4. Wait for any of the following conditions to be true (whichever happens
+        first):
+        1. {{ImageDecoder/[[encoded data]]}} contains enough bytes to
+            completely decode |encodedFrame|.
+        2. {{ImageDecoder/complete}} is `true`.
+        3. {{ImageDecoder/[[fatal error]]}} is `true`.
+    5. If {{ImageDecoder/[[encoded data]]}} does not contain enough bytes to
+        completely decode |encodedFrame|:
+        1. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+        2. Reject |promise| with {{InvalidStateError}} {{DOMException}}.
+        3. Abort these steps.
+    6. Attempt to use {{ImageDecoder/[[codec implementation]]}} to decode
+        |encodedFrame|.
+    7. If decoding produces an error:
+        1. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+        2. Reject |promise| with {{EncodingError}} {{DOMException}}.
+        3. Abort these steps.
+    8. If {{ImageDecoder/[[progressive frame generations]]}} contains an entry
+        keyed by |frameIndex|, remove the entry from the map.
+    9. Let |output| be the decoded image data emitted by
+        {{ImageDecoder/[[codec implementation]]}} corresponding to
+        |encodedFrame|.
+    10. Let |decodeResult| be a new {{ImageDecodeResult}} initialized as
+        follows:
+        1. Assign 'true' to {{ImageDecodeResult/complete}}.
+        2. Assign {{ImageDecodeResult/image}} with the result of running the
+            [=Create a VideoFrame=] algorithm with |output|.
+    13. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+    14. Resolve |promise| with |decodeResult|.
+
+: <dfn for=ImageDecoder>Decode Progressive Frame</dfn> (with |frameIndex| and
+    |promise|)
+:: 1. Assert that {{ImageDecoder/[[tracks established]]}} is `true`.
+    2. Assert that {{ImageDecoder/[[internal selected track index]]}} is not
+        `-1`.
+    3. Let |encodedFrame| be the encoded frame identified by |frameIndex| and
+        {{ImageDecoder/[[internal selected track index]]}}.
+    4. Let |lastFrameGeneration| be `null`.
+    5. If {{ImageDecoder/[[progressive frame generations]]}} contains a map
+        entry with the key |frameIndex|, assign the value of the map entry to
+        |lastFrameGeneration|.
+    6. Wait for any of the following conditions to be true (whichever happens
+        first):
+        1. {{ImageDecoder/[[encoded data]]}} contains enough bytes to decode
+            |encodedFrame| to produce an output who's [=Progressive Image
+            Frame Generation=] does not equal |lastFrameGeneration|.
+        2. {{ImageDecoder/complete}} is `true`.
+        3. {{ImageDecoder/[[fatal error]]}} is `true`.
+    7. If {{ImageDecoder/[[encoded data]]}} does not contain enough bytes to
```

Comment:
I'm not sure this is right, if we reach complete and and there are no more successive frames decoded and there's no error then by definition that frame is the 'complete frame'.

Reply:
We chatted on this, and my previous comment misunderstood your concern. 

> I'm not sure this is right, if we reach complete and and there are no more successive frames decoded and there's no error then by definition that frame is the 'complete frame'.

After chatting, we resolved to a different interpretation. For an output frame, being "complete" means the frame is the full fidelity output (i.e. the highest quality possible for formats that support progressive decoding). This is distinct from our concept of buffered data being "complete" (decoder.complete). 

Here are the scenarios we identified (and their outcomes) for when `decoder.complete = true`:
1. If encoded data is found to be malformed (e.g. the user gave an ArrayBuffer that cuts off in the middle of an encoded frame or the packetization of the encoded frames is mangled), trigger fatal shutdown of the decoder.
2. If we can decode a new successive frame (higher frame generation than previously output), resolve the promise with that frame. Set `result.complete` purely based on whether the output is the full quality output for a progressive image. 
3. if we can't decode a a new successive frame, run the "rejecet infeasible decode algorithm".

For scenario 1, I've added spec steps (not previously handled). I've also made also made the later "decoder produces an error" steps fatal. In short... if the data is identified to be malformed during demuxing or decoing, shut it all down.